src/internet_identity/src/storage.rs

```diff
@@ -396,6 +425,100 @@ impl<M: Memory + Clone> Storage<M> {
+        let num_allocated_buckets: u16 =
+            ((storage_v6.anchor_memory.size() + (BUCKET_SIZE_IN_PAGES as u64) - 1)
+                / BUCKET_SIZE_IN_PAGES as u64) as u16;
+        let mut memory_sizes_in_pages: [u64; mm::MAX_NUM_MEMORIES as usize] =
+            [0u64; mm::MAX_NUM_MEMORIES as usize];
+        memory_sizes_in_pages[ANCHOR_MEMORY_INDEX as usize] = storage_v6.anchor_memory.size();
+        let mm_header = mm::Header {
+            magic: *mm::MAGIC,
+            version: 1u8,
+            num_allocated_buckets,
+            bucket_size_in_pages: BUCKET_SIZE_IN_PAGES,
+            _reserved: [0u8; 32],
+            memory_sizes_in_pages,
+        };
+        let pages_in_allocated_buckets = (BUCKET_SIZE_IN_PAGES * num_allocated_buckets) as u64;
+        memory.grow(pages_in_allocated_buckets - storage_v6.anchor_memory.size());
+        let mm_header_bytes = unsafe {
+            std::slice::from_raw_parts(
+                &mm_header as *const _ as *const u8,
+                std::mem::size_of::<mm::Header>(),
+            )
+        };
+        let mut mm_header_memory = RestrictedMemory::new(memory.clone(), 1..2);
+        let mut writer = Writer::new(&mut mm_header_memory, 0);
+        writer
+            .write_all(mm_header_bytes)
+            .expect("bug: failed to grow memory");
+        // Update bucket-to-memory assignments.
+        // The assignments begin after right after the header, which has the following layout
+        // -------------------------------------------------- <- Address 0
+        // Magic "MGR"                           ↕ 3 bytes
+        // --------------------------------------------------
+        // Layout version                        ↕ 1 byte
+        // --------------------------------------------------
+        // Number of allocated buckets           ↕ 2 bytes
+        // --------------------------------------------------
+        // Bucket size (in pages) = N            ↕ 2 bytes
+        // --------------------------------------------------
+        // Reserved space                        ↕ 32 bytes
+        // --------------------------------------------------
+        // Size of memory 0 (in pages)           ↕ 8 bytes
+        // --------------------------------------------------
+        // Size of memory 1 (in pages)           ↕ 8 bytes
+        // --------------------------------------------------
+        // ...
+        // --------------------------------------------------
+        // Size of memory 254 (in pages)         ↕ 8 bytes
+        // -------------------------------------------------- <- Bucket allocations
+        // ...
+        let buckets_offset: u64 = (3 + 1 + 2 + 2) + 32 + (255 * 8);
+        let mut writer = Writer::new(&mut mm_header_memory, buckets_offset);
+        let mut bucket_to_memory = [mm::UNALLOCATED_BUCKET_MARKER; mm::MAX_NUM_BUCKETS as usize];
+        for i in 0..num_allocated_buckets {
```

Comment:
You could avoid the loop by using `splice`:
```
let mut bucket_to_memory = vec![mm::UNALLOCATED_BUCKET_MARKER; mm::MAX_NUM_BUCKETS as usize];
bucket_to_memory.splice(
    0..num_allocated_buckets as usize,
    vec![0; num_allocated_buckets as usize],
);
```

Requires using vec though, so up to you.

Reply:
Thanks for pointing it out, I'll leave as-is though, ok?